### PR TITLE
Support wildcard/glob in COPY

### PR DESCRIPTION
**Pull request by idodod**
Address #3966

### CODE REVIEW

File: domain/artifact.go [hunk 1]
```diff
@@ -14,6 +14,12 @@ type Artifact struct {
 	Artifact string
 }
 
+// Clone returns a copy of the Artifact
+func (a Artifact) Clone() Artifact {
+	newArtifact := a
```
Comment (mikejholly) [inline domain/artifact.go:19]: Nit: I think `return a` here is equivalent as no pointers involved.

File: earthfile2llb/converter.go [hunk 2]
```diff
@@ -2906,6 +2876,74 @@ func (c *Converter) newCmdID() int {
...
+	}
+
+	if len(targets) == 0 {
+		return nil, errors.Errorf("no matching targets found for pattern %q", parsedTarget.GetLocalPath())
+	}
+
+	return targets, nil
+}
+
+func clonesWithExpandedTargets[T cloneable[T]](expandedTargets []string, c T, setTarget func(t *T, expandedTarget string) error) ([]T, error) {
```
[the remark this answers is not in the document]
Comment (mikejholly) [reply earthfile2llb/converter.go:2934]: Up to you, but I wonder if the use of generics here is more complicated than the simple loop plus `Clone` calls they are meant to abstract.

**Review by mikejholly — APPROVED** [review]
Nice job! Cool new addition. Thanks for adding the docs as well.

**Comment by idodod** [comment]
> Nice job! Cool new addition. Thanks for adding the docs as well.

Thanks, I actually meant and forgot to thank you for paving the way. After your implementation this feature was easy to implement. So thank you for making it easy :)

File: domain/artifact.go (hunk 1 above)
Comment (idodod) [reply domain/artifact.go:19]: good point, it's a leftover from previous commit where I thought I might need to do more to fully clone.
I'll update.

File: earthfile2llb/converter.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (idodod) [reply earthfile2llb/converter.go:2934]: yeah, I was wondering if the indirection is worth it and I'm open to removing it.
I tended to go with generics after all because I liked how it forces the user to clone the object, which I think it might be something that can be easily forgotten. 
I also considered whether this function can be used again, and figured it _might_ in something like `WITH DOCKER --load ./services/*+image`

[the remark this answers is not in the document]
Comment (mikejholly) [reply earthfile2llb/converter.go:2934]: That sounds fair.

**Comment by mikejholly** [comment]
> > Nice job! Cool new addition. Thanks for adding the docs as well.
> 
> Thanks, I actually meant and forgot to thank you for paving the way. After your implementation this feature was easy to implement. So thank you for making it easy :)

Thanks for saying so :smiling_face_with_tear:

File: docs/earthfile/earthfile.md [hunk 3]
```diff
@@ -1100,7 +1163,7 @@ The `WITH DOCKER` clause only supports the command [`RUN`](#run). Other commands
 A typical example of a `WITH DOCKER` clause might be:
 
 ```Dockerfile
-FROM earthly/dind:alpine-3.19-docker-25.0.2-r0
+FROM earthly/dind:alpine-3.19-docker-25.0.3-r2
```
Comment (alexcb) [inline docs/earthfile/earthfile.md:1166]: nit: this seems irrelevant to this PR.

**Review by alexcb — APPROVED** [review]
nice tests too!

File: docs/earthfile/earthfile.md (hunk 3 above)
Comment (alexcb) [reply docs/earthfile/earthfile.md:1166]: ah ha... I see the reason now, nevermind me here :)
